clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -173,20 +198,150 @@ void discoverD2Service() {
+          // worst case, the first read request to this instance will try it again.
+          newReplicasToBeWarmedUp.remove(replica);
+          replicasWarmupInProgressFromPastRefresh.add(replica);
+        } else {
+          futuresToRemove.add(replica);
+        }
+      }
+    }
+    futuresToRemove.forEach(replica -> warmUpInstancesFutures.remove(replica));
+
+    if (replicasWarmupInProgressFromPastRefresh.size() != 0) {
+      LOGGER.info(
+          "H2 connection warmup to {} instances still incomplete from the previous attempt, not retrying again for these instances: {}",
+          replicasWarmupInProgressFromPastRefresh.size(),
+          String.join(",", replicasWarmupInProgressFromPastRefresh));
+    }
+
+    // Remove old instances from warmedUpInstances to keep warmedUpInstances relatively fresh
+    warmedUpInstances.retainAll(newReplicasToBeWarmedUp);
+
+    // Warmup all the new instances
+    newReplicasToBeWarmedUp.removeAll(warmedUpInstances);
+    int instanceNum = newReplicasToBeWarmedUp.size();
+    if (instanceNum == 0) {
+      if (replicasWarmupInProgressFromPastRefresh.size() == 0) {
+        LOGGER.info(
+            "H2 connection warmup to all instances for store {} version {} is completed",
+            storeName,
+            currentFetchedVersion);
+      }
+      return;
+    }
+    LOGGER.info(
+        "Starting H2 connection warm up from fast client to {} instances for store {} version {}",
+        instanceNum,
+        storeName,
+        currentFetchedVersion);
+
+    AtomicInteger numberOfWarmUpSuccess = new AtomicInteger(0);
+    Set<String> warmUpFailedInstances = VeniceConcurrentHashMap.newKeySet();
+    newReplicasToBeWarmedUp.forEach((replica) -> {
+      String url = replica + "/" + QueryAction.HEALTH.toString().toLowerCase();
+      warmUpInstancesFutures.put(replica, r2TransportClient.get(url).whenCompleteAsync((ignore, throwable) -> {
+        if (throwable == null) {
+          warmedUpInstances.add(replica);
+          numberOfWarmUpSuccess.incrementAndGet();
+        } else {
+          warmUpFailedInstances.add(replica);
+        }
+      }, h2ConnWarmupExecutorService));
+    });
+
+    CompletableFuture<Object> warmupResultFuture = new CompletableFuture<>();
+    CompletableFuture.allOf(warmUpInstancesFutures.values().toArray(new CompletableFuture[0]))
+        .whenComplete((response, throwable) -> {
+          if (throwable != null) {
+            LOGGER.error(
+                "H2 connection warmup to {} instances succeeded and {} instances failed. List of failed instances: {}",
+                numberOfWarmUpSuccess.get(),
+                warmUpFailedInstances.size(),
+                String.join(",", warmUpFailedInstances),
+                throwable);
+
+            warmupResultFuture.complete(null);
+          } else {
+            LOGGER.info("H2 connection warmup to {} instances succeeded", numberOfWarmUpSuccess.get());
```

Comment:
I see that we are doing quite a bit of logging here and there regarding to the progress of warmup. Can we standardize the logging to make sure every log contains the necessary info for us to make sense of them? e.g. we are logging about store and version for warmup start but then the error and completion logs no longer have these info.
```
        "Starting H2 connection warm up from fast client to {} instances for store {} version {}",
```

Reply:
Added a common prefix for all logs.